### PR TITLE
Geo: Add coerce support to libs/geo WKT parser

### LINKED ISSUE #43173: SQL: Handling of coerced WKT polygons
WKT parser in SQL cannot handle polygons that were represented in WKT and are not closed, which can be stored if `coerce` flag in mapping is set to `true` but then cannot be retrieved in SQL. That's not an issue for polygons stored in GeoJSON. To reproduce:
```
DELETE test

PUT test 
{
  "mappings": {
    "properties": {
      "shape": {
        "type": "geo_shape",
        "coerce": true
      }
    }
  }
}

PUT test/_doc/1
{
  "shape": "POLYGON ((1 1, -1 1, -1 -1, 1 -1))"
}

POST _sql
{
  "query": "SELECT * FROM test"
}

```

The last command will return an error:

```
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "first and last points of the linear ring must be the same (it must close itself): lats[0]=1.0 lats[3]=-1.0 lons[0]=1.0 lons[3]=1.0"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "first and last points of the linear ring must be the same (it must close itself): lats[0]=1.0 lats[3]=-1.0 lons[0]=1.0 lons[3]=1.0"
  },
  "status": 400
}
```

### PR DESCRIPTION
**Pull request by imotov**
Adds support for coercing not closed polygons and ignoring Z value
to libs/geo WKT parser.

Closes #43173

### CODE REVIEW

File: libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java [hunk 1]
```diff
@@ -52,12 +53,20 @@
...
-    public WellKnownText() {
+    private final boolean coerce;
+    private final boolean ignoreZValue;
 
+    public WellKnownText(boolean coerce, boolean ignoreZValue) {
+        this.coerce = coerce;
+        this.ignoreZValue = ignoreZValue;
+    }
+
+    public WellKnownText() {
```
Comment (talevy) [inline libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java]: is this needed?

File: libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java [hunk 2]
```diff
@@ -408,7 +419,21 @@ private static Polygon parsePolygon(StreamTokenizer stream) throws IOException,
         }
     }
 
-    private static MultiPolygon parseMultiPolygon(StreamTokenizer stream) throws IOException, ParseException {
+    private void closeLinearRingIfCoerced(ArrayList<Double> lats, ArrayList<Double> lons, ArrayList<Double> alts) {
```
Comment (talevy) [inline libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java]: can we add javadocs here to give context to this operation?

**Review by talevy — COMMENTED** [review]
overall looks good! left a few questions and comments

File: libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java (hunk 1 above)
Comment (imotov) [reply libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java]: Yeah, that's probably harmful now. I will remove it.

File: libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java [hunk 3]
```diff
@@ -408,7 +415,25 @@ private static Polygon parsePolygon(StreamTokenizer stream) throws IOException,
         }
     }
 
-    private static MultiPolygon parseMultiPolygon(StreamTokenizer stream) throws IOException, ParseException {
+    /**
+     * Treats supplied arrays as coordinates of a linear ring. If the ring is not closed and coerce is set to true,
+     * the first set of coordinates (lat, lon and alt if available) are added to the end of the arrays.s
```
Comment (talevy) [inline libs/geo/src/main/java/org/elasticsearch/geo/utils/WellKnownText.java]: nit, trailing `.s`

**Review by talevy — APPROVED** [review]

**Pull request merged**